owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/OwnCloudClient.java

```diff
@@ -92,117 +103,55 @@ public void clearCredentials() {
-            method.setRequestHeader(HttpConstants.ACCEPT_ENCODING_HEADER, HttpConstants.ACCEPT_ENCODING_IDENTITY);
-            if (mCredentials.getHeaderAuth() != null && method.getRequestHeader(AUTHORIZATION_HEADER) == null) {
-                method.setRequestHeader(AUTHORIZATION_HEADER, mCredentials.getHeaderAuth());
-            }
-            status = method.execute();
-
-            if (mFollowRedirects) {
-                status = followRedirection(method).getLastStatus();
+        if(mSynchronizeRequests) {
+            synchronized (mRequestMutex) {
+                return saveExecuteHttpMethod(method);
             }
-
-            repeatWithFreshCredentials = checkUnauthorizedAccess(status, repeatCounter);
-            if (repeatWithFreshCredentials) {
-                repeatCounter++;
-            }
-        } while (repeatWithFreshCredentials);
-
-        return status;
+        } else {
+            return saveExecuteHttpMethod(method);
+        }
     }
 
-    private int executeRedirectedHttpMethod(HttpBaseMethod method) throws Exception {
-        boolean repeatWithFreshCredentials;
+    private int saveExecuteHttpMethod(HttpBaseMethod method) throws Exception {
         int repeatCounter = 0;
         int status;
 
+        if(mFollowRedirects) {
+            method.setFollowRedirects(true);
+        }
+
+        boolean retry;
         do {
+            repeatCounter++;
+            retry = false;
             String requestId = RandomUtils.generateRandomUUID();
 
             // Header to allow tracing requests in apache and ownCloud logs
             Timber.d("Executing in request with id %s", requestId);
-            method.setRequestHeader(OC_X_REQUEST_ID, requestId);
+            method.setRequestHeader(HttpConstants.OC_X_REQUEST_ID, requestId);
             method.setRequestHeader(HttpConstants.USER_AGENT_HEADER, SingleSessionManager.getUserAgent());
             method.setRequestHeader(HttpConstants.ACCEPT_ENCODING_HEADER, HttpConstants.ACCEPT_ENCODING_IDENTITY);
-            if (mCredentials.getHeaderAuth() != null) {
+            if (mCredentials.getHeaderAuth() != null && !mCredentials.getHeaderAuth().isEmpty()) {
                 method.setRequestHeader(AUTHORIZATION_HEADER, mCredentials.getHeaderAuth());
             }
+
             status = method.execute();
 
-            repeatWithFreshCredentials = checkUnauthorizedAccess(status, repeatCounter);
-            if (repeatWithFreshCredentials) {
-                repeatCounter++;
+            if (!mFollowRedirects &&
+                    !method.getFollowRedirects() &&
+                    mConnectionValidator != null &&
+                    (status == HttpConstants.HTTP_MOVED_TEMPORARILY ||
+                            (!(mCredentials instanceof OwnCloudAnonymousCredentials) &&
+                                    status == HttpConstants.HTTP_UNAUTHORIZED))) {
```

Comment:
I would try to simplify or move this to a function. It's kind of hard to follow the conditions 🤔